wfl/autoparallelize/base.py

```diff
@@ -9,45 +9,133 @@
 from .pool import do_in_pool
 from .remote import do_remotely
 
+# NOTES from a previous implementation that does not work with sphinx docstring parsing
+# may not be exactly correct, but can definitely be made to work (except sphinx)
+#
+# note that it is possible to define iloop so that the decorator functionality can be
+# done with
+#
+#     parallelized_op = iloop(op, "parallelized_op", "Atoms",. [iloop_arg_1 = ... ])
+#
+# this is done by renaming the current "iloop" to "_iloop_wrapper", and defining a new "iloop"
+# as something like
+#
+#     def iloop(op, op_name, input_iterator_contents, **iloop_kwargs):
+#         f = functools.partial(_iloop_wrapper, op, **iloop_kwargs)
+#         f.__name__ = op_name
+#         f.__doc__ = iloop_docstring(op.__doc__, input_iterator_contents)
+#         return f
+#
+# it also requires that the code parsing the stack trace to match up the RemoteInfo dict
+# detects the presence of "iloop" in the stack trace, and instead uses
+#
+#     inspect.getfile(op) + "::" + op.__name__
+#
+# instead of the stack trace file and function
+#
+# sphinx docstring parsing cannot handle this, though, because the module associated with
+# the "parallelized_op" symbol is that of "iloop", rather than the python file in which
+# it is actually defined.  As a result, sphinx assumes it's just some imported symbol and does
+# not include its docstring.  There does not appear to be any way to override that on a
+# per-symbol basis in current sphinx versions.
+
+_iloop_docstring_params_pre = (
+"""inputs: iterable({iterable_type})
+    input quantities of type {iterable_type}
+outputs: OutputSpec or None
+    where to write output atomic configs, or None for no output (i.e. only side-effects)
+""")
+
+_iloop_docstring_params_post = (
+"""ITERABLE_LOOP_RELATED:
+
+    - num_python_subprocesses: int, default os.environ['WFL_NUM_PYTHON_SUBPROCESSES']
```

Comment:
Sometime go we talked about making a class (like `RemoteInfo`) to gather all of the autoparallelization arguments in one place (issue #65), so it's clearer what arguments go with the function parallelised and and we can modify them without modifying the signature of the auto-parallelised function. Is this a good place to add that? Not sure how much work that'd be; I can also add this to a subsequent PR.

Reply:
Now that it's an iterator, it's a bit cleaner than it used to be - at least you don't have to remember to add all the arguments when you define the wrapped function.  It would clean up the docstrings - right now every wrapped function will have the "ITERABLE_LOOP related" section. Maybe that means this is a good time to do this.  Let me try to code something up.  I'll let you know when there's something to review.

Reply:
What about calling it `AutoparaInfo`, by analogy to `RemoteInfo`?

Reply:
Yup that sounds good, thanks!

Reply:
Somehow this ended up as a much harder software engineering task than I envisioned.  Almost ready now.

Reply:
Updated now. @gelzinyte let me know what you think - you can see how `wfl/calculators/generic.py::run()` does the wrapping, and how `tests/test_autoparallelize.py::test_pool_speedup()` passes info at script calling time.

Reply:
This looks good, thank you for the update. I think it will make adding new functions & using parallelization much more intuitive.